src/client/renderer/r_bsp_light.c

```diff
@@ -322,3 +325,292 @@ void R_DrawBspLights(void) {
+						const vec_t blend = ((color[i]) * stain_alpha) + ((pfBL[i] / 255.0) * dst_stain_alpha);
+						const uint8_t new_color = (uint8_t) (Clamp(blend, 0.0, 1.0) * 255.0);
+
+						if (pfBL[i] != new_color) {
+							pfBL[i] = new_color;
+							changes = true;
+							(*num)++;
+						}
+					}
+				}
+			}
+
+			if (!changes) {
+				continue;
+			}
+
+			surf->stainmap_dirty = true;
+
+			R_BindDiffuseTexture(surf->stainmap->texnum);
+			glTexSubImage2D(GL_TEXTURE_2D, 0, surf->lightmap_s, surf->lightmap_t, smax, tmax, GL_RGB, GL_UNSIGNED_BYTE,
+							surf->stainmap_buffer);
+			R_GetError("stain");
+		}
+
+		g_array_append_vals(node_list, &node->children, 2);
+	}
+}
+
+
+		/*
+static void R_RotateLightsForBspInlineModel(const r_entity_t *e) {
+	static vec3_t light_origins[MAX_LIGHTS];
+	static int16_t frame;
+
+	// for each frame, backup the light origins
+	if (frame != r_locals.frame) {
+		for (uint16_t i = 0; i < r_view.num_lights; i++) {
+			VectorCopy(r_view.lights[i].origin, light_origins[i]);
+		}
+		frame = r_locals.frame;
+	}
+
+	// for malformed inline models, simply return
+	if (e && e->model->bsp_inline->head_node == -1) {
+		return;
+	}
+
+	// for well-formed models, iterate the lights, transforming them into model
+	// space and marking surfaces, or restoring them if the model is NULL
+
+	const r_bsp_node_t *nodes = r_model_state.world->bsp->nodes;
+	r_light_t *l = r_view.lights;
+
+	for (uint16_t i = 0; i < r_view.num_lights; i++, l++) {
+		if (e) {
+			Matrix4x4_Transform(&e->inverse_matrix, light_origins[i], l->origin);
+			R_MarkLight(l, nodes + e->model->bsp_inline->head_node);
+		} else {
+			VectorCopy(light_origins[i], l->origin);
+		}
+	}
+}
+		*/
+
+/**
+ * @brief Add a stain to the map.
+ */
+uint32_t R_AddStain(const vec3_t org, const vec4_t color, const vec_t size) {
+
+	if (!r_stain_map->integer) {
+		return 0;
+	}
+
+	uint32_t num = 0;
+
+	for (int32_t i = 0; i < r_model_state.world->bsp->num_inline_models; ++i) {
+		
+		r_bsp_inline_model_t *model = r_model_state.world->bsp->inline_models + i;
+		vec3_t origin;
+
+		// bad nodes
+		if (model->head_node == -1 ||
+			!model->num_surfaces) {
+			continue;
+		}
+
+		r_bsp_node_t *node = r_model_state.world->bsp->nodes + model->head_node;
+
+		cl_entity_t *entity = NULL;
+
+		// find the entity that owns us
+		if (i != 0) {
+
+			for (uint16_t x = 0; x < cl.frame.num_entities; x++) {
+
+				const uint32_t snum = (cl.frame.entity_state + x) & ENTITY_STATE_MASK;
+				const entity_state_t *s = &cl.entity_states[snum];
+
+				if (s->solid < SOLID_BSP) {
+					continue;
+				}
+				
+				const cm_bsp_model_t *mod = cl.cm_models[s->model1];
+
+				if (mod->head_node == node->model->bsp_inline->head_node) {
+					entity = &cl.entities[s->number];
+					break;
+				}
+			}
+
+			if (!entity) {
+				continue;
+			}
+
+			Matrix4x4_Transform(&entity->inverse_matrix, org, origin);
+		} else {
+			VectorCopy(org, origin);
+		}
+
+		R_StainNode(origin, color, size, node, &num);
```

Comment:
It'd be cleaner and more efficient to loop once over `cl.frame.entities` and deal with ones that reference BSP models, than to iterate the BSP models and then iterate the entities in loop.